src/plugins/oidc/login.ts

```diff
@@ -57,70 +44,142 @@ const authorize = async (org: OrgData, scope: string) => {
+  org: OrgData,
+  scope: string,
+  urls: () => { deviceAuthorizationUrl: string; tokenUrl: string }
+) => {
+  const { deviceAuthorizationUrl, tokenUrl } = urls();
+
+  const buildOidcAuthorizeRequest = () => {
+    if (org.providerType === undefined) {
+      throw "Login requires a configured provider type.";
+    }
+
+    validateProviderDomain(org);
+    // This is the "org" authorization server; the okta.apps.* scopes are not
+    // available with custom authorization servers
+    return {
+      init: {
+        method: "POST",
+        headers: OIDC_HEADERS,
+        body: urlEncode({
+          client_id: org.clientId,
+          scope,
+        }),
+      },
+      url: deviceAuthorizationUrl,
+    };
+  };
+  const buildOidcTokenRequest = (authorize: AuthorizeResponse) => {
+    if (org.providerType === undefined) {
+      throw "Login requires a configured provider type.";
+    }
+    validateProviderDomain(org);
+
+    return {
+      url: tokenUrl,
+      init: {
+        method: "POST",
+        headers: OIDC_HEADERS,
+        body: urlEncode({
+          client_id: org.clientId,
+          device_code: authorize.device_code,
+          grant_type: DEVICE_GRANT_TYPE,
+        }),
+      },
+    };
+  };
+  return {
+    providerType: providerType(org),
+    validateResponse,
+    buildAuthorizeRequest: buildOidcAuthorizeRequest,
+    buildTokenRequest: buildOidcTokenRequest,
+    processAuthzExpiry: (authorize) => ({
+      expires_in: authorize.expires_in,
+      interval: authorize.interval,
+    }),
+    processAuthzResponse: (authorize) => ({
+      user_code: authorize.user_code,
+      verification_uri_complete: authorize.verification_uri_complete,
+    }),
+  } as OidcLoginStepHelpers<AuthorizeResponse>;
+};
+
 /** Logs in to an Identity Provider via OIDC */
-export const oidcLogin = async (org: OrgData, scope: string) => {
-  if (org.providerType === undefined) {
-    throw "Login requires a configured provider type.";
-  }
-  const authorizeResponse = await authorize(org, scope);
+export const oidcLogin = async <A, T>(context: OidcLoginStepHelpers<A>) => {
+  const {
+    providerType,
+    buildAuthorizeRequest,
+    buildTokenRequest,
+    processAuthzExpiry,
+    processAuthzResponse,
+    validateResponse,
+  } = context;
+  const deviceAuthorizationResponse = await authorize<A>(
+    buildAuthorizeRequest(),
+    validateResponse
+  );
+  const { user_code, verification_uri_complete } = processAuthzResponse(
+    deviceAuthorizationResponse
+  );
   print2(`Please use the opened browser window to continue your P0 login.
-  
-When prompted, confirm that ${capitalize(org.providerType)} displays this code:
-
-  ${authorizeResponse.user_code}
 
-Waiting for authorization...
-`);
-  void open(authorizeResponse.verification_uri_complete);
-  const oidcResponse = await waitForActivation(org, authorizeResponse);
-  return oidcResponse;
+    When prompted, confirm that ${capitalize(providerType)} displays this code:
```

Comment:
We probably want a mapping of supported provider types to human-friendly labels. E.g., I think the P0-internal `providerType` for "Entra ID" is `azure-ad`; merely capitalizing this will likely confuse the end user.

Reply:
makes sense